`error CS0103: The name '_default_T' does not exist in the current context`

I'm not sure if this is me missing something in my C# code, or a bug in Dafny?

```
$ dafny /compile:1  types.dfy concrete_types.cs                  

Dafny program verifier finished with 0 verified, 0 errors
Wrote textual form of target program to types.cs
Errors compiling program into types
(1780,51): error CS0103: The name '_default_T' does not exist in the current context
```

Adding `/spillTargetCode:1` shows that indeed the generated code uses `_default_T` without declaring it.

File types.dfy:
```dafny
module {:extern "M"} M {
    type {:extern "T"} T
    datatype D = D (
        T: T
    )
}
```

File concrete_types.cs:
```csharp
namespace M {
    public class T {
    }
}
```

--- a/Source/Dafny/Compiler-Csharp.cs
+++ b/Source/Dafny/Compiler-Csharp.cs
@@ -44,7 +44,20 @@ namespace Microsoft.Dafny
     static string FormatTypeDescriptorVariable(string typeVarName) => $"_td_{typeVarName}";
     static string FormatTypeDescriptorVariable(TypeParameter tp) => FormatTypeDescriptorVariable(tp.CompileName);
     const string TypeDescriptorMethodName = "_TypeDescriptor";
-    static string FormatDefaultTypeParameterValue(TypeParameter tp) => $"_default_{tp.CompileName}";
+    static string FormatDefaultTypeParameterValue(TypeParameter tp) {
+      if (tp is OpaqueType_AsParameter) {
+        // This is unusual. Typically, the compiler never needs to compile an opaque type, but this opaque type
+        // is apparently an :extern. It's difficult to say what the compiler could do in this situation, since
+        // it doesn't know how to generate code that produces a legal value of the opaque type. If we don't do
+        // anything different from the common case (the "else" branch below), then the code emitted will not
+        // compile (see github issue #1151). So, to do something a wee bit better, we emit a placebo value. This
+        // will only work when the opaque type is in the same module and has no type parameters.
+        return $"default({tp.CompileName})";
+      } else {
+        // this is the common case
+        return $"_default_{tp.CompileName}";
+      }
+    }
 
     protected override void EmitHeader(Program program, TargetWriter wr) {
       wr.WriteLine("// Dafny program {0} compiled into C#", program.Name);

--- a/Test/git-issues/git-issue-1151-concrete.cs
+++ b/Test/git-issues/git-issue-1151-concrete.cs
@@ -1,0 +1,24 @@
+namespace M {
+  // try an unusual type here, namely a struct
+  public struct T {
+  }
+
+  public partial class __default {
+    public static T GetT() {
+      return new T();
+    }
+  }
+
+  // By declaring U a class here, it allows "null" as a value.
+  // In other words, the opaque type M.U declared in the Dafny program
+  // evidently stands for what in Dafny could have been declared as
+  // type U? for a class U.
+  public class U {
+  }
+
+  public struct V {
+  }
+
+  public struct W<A> {
+  }
+}

--- a/Test/git-issues/git-issue-1151.dfy
+++ b/Test/git-issues/git-issue-1151.dfy
@@ -1,0 +1,29 @@
+// RUN: %dafny /compile:3 "%s" git-issue-1151-concrete.cs > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+module {:extern "M"} M {
+  type {:extern "T"} T
+  method {:extern} GetT() returns (t: T)
+
+  datatype D = D(t: T)
+  datatype E<X> = E(t: T, x: X)
+
+  type {:extern "U"} U(0)
+  type {:extern "V"} V(0)
+  datatype F = F(u: U, v: V)
+}
+
+method Main() {
+  var t := M.GetT();
+
+  var d: M.D;
+  d := M.D(t);
+  print d, "\n";
+
+  var e: M.E<int>;
+  e := M.E(t, 45);
+  print e, "\n";
+
+  var f: M.F;
+  print f, "\n";
+}

--- a/Test/git-issues/git-issue-1151.dfy.expect
+++ b/Test/git-issues/git-issue-1151.dfy.expect
@@ -1,0 +1,5 @@
+
+Dafny program verifier finished with 1 verified, 0 errors
+M_Compile.D.D(M.T)
+M_Compile.E.E(M.T, 45)
+M_Compile.F.F(null, M.V)

